purchases/src/test/java/com/revenuecat/purchases/attributes/SubscriberAttributesBackendTests.kt

```diff
@@ -27,125 +34,217 @@ class SubscriberAttributesBackendTests {
             appUserID,
-            {
-                received = true
-            },
-            { _, _ ->
-                fail("should be success")
-            }
+            expectedOnSuccess,
+            unexpectedOnError
         )
-        assertThat(received).isTrue()
+        assertThat(receivedOnSuccess).isTrue()
     }
 
     @Test
     fun `posting null subscriber attributes works`() {
         mockResponse("/subscribers/$appUserID/attributes", 200)
 
-        var received = false
         underTest.postSubscriberAttributes(
             mapOf("email" to SubscriberAttribute("email", null)),
             appUserID,
-            {
-                received = true
-            },
-            { _,  _ ->
-                fail("should be success")
-            }
+            expectedOnSuccess,
+            unexpectedOnError
         )
-        assertThat(received).isTrue()
+        assertThat(receivedOnSuccess).isTrue()
     }
 
     @Test
     fun `error when posting attributes`() {
         mockResponse("/subscribers/$appUserID/attributes", 0, clientException = IOException())
 
-        var receivedError: PurchasesError? = null
-        var receivedSyncedSuccessfully: Boolean? = null
         underTest.postSubscriberAttributes(
             mapOf("email" to SubscriberAttribute("email", null)),
             appUserID,
-            {
-                fail("should be failure")
-            },
-            { error,  syncedSuccessfully ->
-                receivedError = error
-                receivedSyncedSuccessfully = syncedSuccessfully
-            }
+            unexpectedOnSuccess,
+            expectedOnError
         )
+
         assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.NetworkError)
         assertThat(receivedSyncedSuccessfully).isFalse()
+        assertThat(receivedAttributeErrors).isEmpty()
+    }
+
+    @Test
+    fun `attributes validation error when posting attributes`() {
+        mockResponse(
+            "/subscribers/$appUserID/attributes",
+            400,
+            expectedResultBody = "{" +
+                "'code': 7262," +
+                "'message': 'Some subscriber attributes keys were unable to saved.'," +
+                "'attribute_errors':" +
+                "[{'key_name': 'email', 'message': 'Value is not a valid email address.'}]}"
+        )
+
+        underTest.postSubscriberAttributes(
+            mapOf("email" to SubscriberAttribute("email", null)),
+            appUserID,
+            unexpectedOnSuccess,
+            expectedOnError
+        )
+
+        assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.InvalidSubscriberAttributesError)
+        assertThat(receivedSyncedSuccessfully).isTrue()
+        assertThat(receivedAttributeErrors!!.size).isEqualTo(1)
+        assertThat(receivedAttributeErrors!![0].keyName).isEqualTo("email")
+        assertThat(receivedAttributeErrors!![0].message).isEqualTo("Value is not a valid email address.")
```

Comment:
are `!!`s necessary here? if the inner exception evaluates to null, it won't be equalTo the actual value anyway, right?

Reply:
I need to do either `!!` or `?` since the value is nullable. Doing `!!` guarantees that if the value is null, which I don't expect, the test will fail

Reply:
what I meant is if you do `?`, won't the assertion fail anyway? 
`assertThat(null).isEqualTo("email")` will fail, right? 
but if you have multiple statements like: 
```kotlin
assertThat(nullThing!!).isEqualTo("email")
assertThat(somethingTrue).isFalse()
```
this will throw on the first one and you won't see what happened on the second one. and the error you get will just say that receivedAttributesErrors failed to unwrap. To figure out what happened, you'd go to the line number and see where it crashed.

by using `?`, however: 

```kotlin
assertThat(nullThing?).isEqualTo("email")
assertThat(somethingTrue).isFalse()
```
Would evaluate both, say that the first assertion failed because you expected null == email, and that the second one expected somethingTrue to be false, but got true.